api/turing/api/pod_log_api.go

```diff
@@ -48,78 +68,217 @@ func (c PodLogController) ListPodLogs(_ *http.Request, vars RequestVars, _ inter
 		}
 	}
 
-	var componentType string
-	switch componentType, _ = vars.get("component_type"); componentType {
-	case servicebuilder.ComponentTypes.Router,
-		servicebuilder.ComponentTypes.Enricher,
-		servicebuilder.ComponentTypes.Ensembler:
-	case "":
-		componentType = "router"
-	default:
-		return BadRequest(fmt.Sprintf("Invalid component type '%s'", componentType),
-			"must be one of router, enricher or ensembler")
+	componentType, err := getComponentType(servicebuilder.ComponentTypes.Router, vars, routerAllowedComponents)
+	if err != nil {
+		return BadRequest(err.Error(), "must be one of router, enricher or ensembler")
```

Comment:
Can this `"must be one of router, enricher or ensembler"` message be generated based on `routerAllowedComponents`, instead of a hard-coded string?

Reply:
Also, a better idea would be to automate this by specifying the schema of query variables together with validation tags and then use `ParseVars` to parse and validate them. 

Example: https://github.com/gojek/turing/blob/9a728b1132aa2d7f7b1a4af093d103fcf5ba5afe/api/turing/api/ensembling_job_api.go#L97

Reply:
I hesitated on this while implementing the logic because, 

> > > > There are a couple of changes from the router logs endpoint. Here's a comparison: ...
> > > 
> > > 
> > > How this `logging_url` would work for multiple executor pods? I.e. If I have a job with 5 executor replica, then I'd expect to have my log entries to be grouped by the `component_type` (i.e `executor`) and then `logging_url` should be common for each executor pod
> > 
> > 
> > It works exactly as how the current router logs work since the logic is shared. So if a call is `GET ​/projects​/{project_id}​/jobs​/{job_id}​/logs?component_type=executor` then it gets all the logs from all the pods with the same logic as turing routers.
> > It will look something like this, the logging URL will be common which is why it is outside the log list.
> > ...
> 
> sorry @ashwinath, I somehow missed this comment. So my concern and the reason of why I left that comment is that in this case `loggingURLTemplate` should support multiple pods simultaneously i.e. if `component_type` is `executor` and there are 3 replicas of executor (pod1, pod2, pod3), then logging URL should be something like `http://www.example.com/logging?pod_name=pod1&pod_name=pod2&pod_name=pod3&namespace=world"`. Or alternatively, it should support regex or something

What is the use case for this? I find this to be very oddly specific and not sure how this can be used in the UI. All podnames are too similar. I don't see the value of specifying the pod name. Let me give you an example of how the pod names are like:

Builder: only 1 pod so there is no differentiation `component_type=image_builder` is good enough.
Driver: only 1 driver so there is no differentiation `component_type=driver` is good enough.

Executor: may have multiple but `component_type=executor` will filter all of them out. So what is left is to differentiate between the pods. From what I remember, all executor pod have names like this `<ensembler_name>-<replicaset-hash>-exec-<replica number indexed from 0>`. Personally I don't see any value differentiating pod 1 from 2 and 3.

Reply:
I don't think this can be done with the current playground validator. I can't put a default value but i need it to be `oneof=router enricher ensembler`. But if it's empty it should be router. I'm not sure if the mux schema can do this or if validator can do this.

Reply:
Do you remember the reason why David wrote it like this in the first place? Instead of using playground validator.

Reply:
Yeah, the `default` tag was only added recently and is still not merged into the main branch: https://github.com/gorilla/schema/pull/183, but as discussed separately, maybe we don't really need default values at all.

> Do you remember the reason why David wrote it like this in the first place? 
Can't remember reviewing it :)

Reply:
As discussed in private, we will remove default values and enforce that `component_type` will be required, no empty query params will be allowed (i.e. `?head_lines=&tail_lines=&component_type=router`). UI changes are expected in a separate PR that will be picked up by @romanwozniak.

Reply:
I have overhauled the parser/validator.